### PR TITLE
Fix DCE of begin_borrow with @guaranteed operand

### PR DESCRIPTION
**Pull request by meg-gupta**
For the lower most begin_borrows with @guaranteed operands, there was no @guaranteed phi dependency computation, restructure the code and fix this case.

### CODE REVIEW

**Comment by meg-gupta** [comment]
@swift-ci smoke test

**Review by atrick — APPROVED** [review]
Thanks for figuring this out. It's beyond my ability to fully comprehend. I'm hoping that an extended liveness abstraction can avoid much of the complexity.

File: lib/SILOptimizer/Transforms/DeadCodeElimination.cpp [hunk 1]
```diff
@@ -283,37 +283,33 @@ void DCE::markLive() {
...
         auto *borrowInst = cast<BeginBorrowInst>(&I);
+        // Populate guaranteedPhiDependencies for this borrowInst
+        findGuaranteedPhiDependencies(BorrowedValue(borrowInst));
+        auto disableBorrowDCE = [&](SILValue borrow) {
+          visitTransitiveEndBorrows(borrow, [&](EndBorrowInst *endBorrow) {
+            markInstructionLive(endBorrow);
+          });
+        };
+        // If we have a begin_borrow of a @guaranteed operand, disable DCE'ing
+        // of parent borrow scopes. Dead reborrows needs complex handling, whuch
```
Comment (meg-gupta) [inline lib/SILOptimizer/Transforms/DeadCodeElimination.cpp:295]: TODO: Piggyback the typo in another PR

**Review by nate-chandler — COMMENTED** [review]
LGTM

**Pull request merged**